homeassistant/components/mqtt/water_heater.py

```diff
@@ -0,0 +1,438 @@
+            except ValueError:
+                _LOGGER.error("Could not parse temperature from %s", payload)
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def handle_current_temperature_received(msg: ReceiveMessage) -> None:
+            """Handle current temperature coming via MQTT."""
+            handle_temperature_received(
+                msg, CONF_CURRENT_TEMP_TEMPLATE, "_attr_current_temperature"
+            )
+
+        add_subscription(
+            topics, CONF_CURRENT_TEMP_TOPIC, handle_current_temperature_received
+        )
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def handle_target_temperature_received(msg: ReceiveMessage) -> None:
+            """Handle target temperature coming via MQTT."""
+            handle_temperature_received(
+                msg, CONF_TEMP_STATE_TEMPLATE, "_attr_target_temperature"
+            )
+
+        add_subscription(
+            topics, CONF_TEMP_STATE_TOPIC, handle_target_temperature_received
+        )
+
+        @callback
+        def handle_mode_received(
+            msg: ReceiveMessage, template_name: str, attr: str, mode_list: str
+        ) -> None:
+            """Handle receiving listed mode via MQTT."""
+            payload = render_template(msg, template_name)
+
+            if payload not in self._config[mode_list]:
+                _LOGGER.error("Invalid %s mode: %s", mode_list, payload)
+            else:
+                setattr(self, attr, payload)
+                get_mqtt_data(self.hass).state_write_requests.write_state_request(self)
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def handle_current_mode_received(msg: ReceiveMessage) -> None:
+            """Handle receiving mode via MQTT."""
+            handle_mode_received(
+                msg,
+                CONF_MODE_STATE_TEMPLATE,
+                "_attr_current_operation",
+                CONF_MODE_LIST,
+            )
+
+        add_subscription(topics, CONF_MODE_STATE_TOPIC, handle_current_mode_received)
+
+        self._sub_state = subscription.async_prepare_subscribe_topics(
+            self.hass, self._sub_state, topics
+        )
+
+    async def _subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        await subscription.async_subscribe_topics(self.hass, self._sub_state)
+
+    async def _publish(self, topic: str, payload: PublishPayloadType) -> None:
+        if self._topic[topic] is not None:
+            await self.async_publish(
+                self._topic[topic],
+                payload,
+                self._config[CONF_QOS],
+                self._config[CONF_RETAIN],
+                self._config[CONF_ENCODING],
+            )
+
+    async def _set_temperature_attribute(
+        self,
+        temp: float | None,
+        cmnd_topic: str,
+        cmnd_template: str,
+        state_topic: str,
+        attr: str,
+    ) -> bool:
```

Comment:
This is part of `MqttTemperatureControlEntity` and available as a class method, so it can be removed here.